w1r3/go/main.go

```diff
@@ -0,0 +1,582 @@
+				append([]attribute.KeyValue{attribute.String("ssb.op", uploader.name)}, commonAttributes...)...))
+
+		upload_start := time.Now()
+		objectHandle, err := uploader.uploader(uploadContext, transport.client, config.bucketName, objectName, data[0:objectSize])
+		if err != nil {
+			uploadSpan.SetStatus(codes.Error, "error during upload")
+			uploadSpan.RecordError(err)
+			uploadSpan.End()
+			span.End()
+			continue
+		}
+		duration := time.Since(upload_start)
+		histogram.Record(uploadContext, duration.Milliseconds(), metric.WithAttributes(
+			append([]attribute.KeyValue{attribute.String("ssb.op", uploader.name)}, commonAttributes...)...))
+		uploadSpan.End()
+
+		discard := make([]byte, 2*MiB)
+		for r := range 3 {
+			op := fmt.Sprintf("READ[%d]", r)
+			downloadContext, downloadSpan := tracer.Start(
+				spanContext, "ssb::download", trace.WithAttributes(
+					append([]attribute.KeyValue{attribute.String("ssb.op", op)}, commonAttributes...)...))
+
+			download_start := time.Now()
+			objectReader, err := objectHandle.NewReader(downloadContext)
+			if err != nil {
+				log.Printf("Error in ssb::upload %d transport %s size %d: %v", i, transport.name, objectSize, err)
+				downloadSpan.SetStatus(codes.Error, "error while opening reader")
+				downloadSpan.RecordError(err)
+				downloadSpan.End()
+				continue
+			}
+			for {
+				_, err := objectReader.Read(discard)
+				if err != nil {
+					break
+				}
+			}
+			err = objectReader.Close()
+			if err != nil {
+				downloadSpan.SetStatus(codes.Error, "error while closing reader")
+				downloadSpan.RecordError(err)
+			}
+			downloadSpan.End()
+			duration := time.Since(download_start)
+			histogram.Record(downloadContext, duration.Milliseconds(), metric.WithAttributes(
+				append([]attribute.KeyValue{attribute.String("ssb.op", op)}, commonAttributes...)...))
+		}
+
+		d := objectHandle.Retryer(storage.WithPolicy(storage.RetryAlways))
+		d.Delete(spanContext)
+		span.End()
+	}
+}
+
+type Uploader struct {
+	name     string
+	uploader func(ctx context.Context, client *storage.Client, bucketName string, objectName string, data []byte) (*storage.ObjectHandle, error)
+}
+
+func singleShotUpload(ctx context.Context, client *storage.Client, bucketName string, objectName string, data []byte) (*storage.ObjectHandle, error) {
+	bucket := client.Bucket(bucketName)
+	o := bucket.Object(objectName)
+	objectWriter := o.If(storage.Conditions{DoesNotExist: true}).NewWriter(ctx)
+	objectWriter.ChunkSize = len(data) + 1
+	if _, err := io.Copy(objectWriter, bytes.NewBuffer(data)); err != nil {
+		return o, err
+	}
+	return o, objectWriter.Close()
+}
+
+func resumableUpload(ctx context.Context, client *storage.Client, bucketName string, objectName string, data []byte) (*storage.ObjectHandle, error) {
+	bucket := client.Bucket(bucketName)
+	o := bucket.Object(objectName)
+	objectWriter := o.If(storage.Conditions{DoesNotExist: true}).NewWriter(ctx)
+	objectWriter.ChunkSize = 2 * MiB
```

Comment:
So is this sufficient to force resumable for smaller files? Can you add a comment?

Reply:
Done.